src/GraphQL.php

```diff
@@ -70,6 +70,9 @@ public function schema($schema = null): Schema
         $schemaQuery = $schema['query'] ?? [];
         $schemaMutation = $schema['mutation'] ?? [];
         $schemaSubscription = $schema['subscription'] ?? [];
+        $schemaTypes = $schema['types'] ?? [];
+
+        $this->addTypes($schemaTypes);
```

Comment:
This one is interesting.

Every time a query is now run, it adds the schema types to the internal `types` list.

Previously this was only done once when constructing _the_ GraphQL instance.

Multiple invocation within the same process of the same query will work correctly, as `types` are stored by their name.

Now, if multiple schemas with same named types are used... they will overwrite previous one but probably "just" work. I guess.

It's in fact similar to `typesInstances` with the difference that they're **cleared** before every query, so instances of different schemas are never mixed.

I still have to think about the impact here but if you've any insights here, please share them!

Reply:
@mfn the thing is this whole mechanism seems to rely on php's request lifetime already, and I've not addressed that problem to reach my goal here – really we should do better, but it looks like a significant piece of work to get there I think. The `GraphQL::type(...)` helper would need to become schema-aware, and know to look within a separate pool of independently built/cached types depending on the current schema in use.

I figured the tests would exercise this problem and break for this reason, but they didn't, probably because even when >1 queries are executed within the same test, they're happening against the same schema, the same as in a request. Since a request can only be performed against one schema at a time, I figured this solution would probably do for now, but really I'd prefer this scenario was handled better.